### PR TITLE
add ending slashes to paths

### PR DESCRIPTION
**Pull request by abelcastilloavant**
Extending the idea to add ending slashes to paths in `s3read` to `s3store` and `s3exists`, via a helper function.

### CODE REVIEW

**Comment by peterhurford** [comment]
looks good, thanks! 👍 Can you version?

File: tests/testthat/test-s3read.R [hunk 1]
```diff
@@ -16,6 +16,7 @@ local({
       expect_error(s3read("key", cache = NA, serialize = FALSE))
     })  
     test_that("if cache is TRUE, it does not error", {
+      browser()
```
Comment (peterhurford) [inline tests/testthat/test-s3read.R]: No browsers

Comment (abelcastilloavant) [inline tests/testthat/test-s3read.R]: That must be why Travis passed, but a test failed locally. :)

File: tests/testthat/test-s3read.R [hunk 2]
```diff
@@ -16,7 +16,7 @@ local({
       expect_error(s3read("key", cache = NA, serialize = FALSE))
     })  
     test_that("if cache is TRUE, it does not error", {
-      expect_equal(s3read("key", cache = TRUE, serialize = FALSE), "value")
+      expect_equal(s3read("key", cache = TRUE, serialize = FALSE), TRUE)
```
Comment (peterhurford) [inline tests/testthat/test-s3read.R]: What happened here?

Comment (abelcastilloavant) [inline tests/testthat/test-s3read.R]: The way the mocks are set up, https://github.com/robertzk/s3mpi/blob/master/R/s3read.r#L43 writes `TRUE` into `value`. Is this the desired behavior?

Comment (peterhurford) [inline tests/testthat/test-s3read.R]: Looks right -- curious why tests passed before. @robertzk ?

Comment (peterhurford) [inline tests/testthat/test-s3read.R]: Notably this test fails now.

Comment (abelcastilloavant) [inline tests/testthat/test-s3read.R]:
```
debug at .../s3mpi/R/s3read.r#36: if (!isTRUE(cache) || is.null(getOption("s3mpi.cache"))) {
    value <- s3.get(s3key, ...)
} else if (is.not_cached(value <- s3cache(s3key))) {
    value <- s3.get(s3key, ...)
    s3cache(s3key, value)
}
Browse[3]> ls()
[1] "cache"     "name"      "path"      "s3key"     "serialize"
Browse[3]> n
debug at .../s3mpi/R/s3read.r#36: if (is.not_cached(value <- s3cache(s3key))) {
    value <- s3.get(s3key, ...)
    s3cache(s3key, value)
}
Browse[3]> ls()
[1] "cache"     "name"      "path"      "s3key"     "serialize"
Browse[3]> (is.not_cached(s3cache(s3key)))
[1] FALSE
Browse[3]> n
debug at .../s3mpi/R/s3read.r#44: if (isTRUE(serialize)) {
    s3normalize(value, TRUE)
} else {
    value
}
Browse[3]> value
[1] TRUE
```

**Comment by peterhurford** [comment]
LGTM if tests pass

**Comment by abelcastilloavant** [comment]
Test failure at commit 1c4b99f :
```
1. Failure: if cache is TRUE, it does not error
  s3read("key", cache = TRUE, serialize = FALSE) not equal to TRUE.
  Types not compatible: character vs logical
```

**Comment by abelcastilloavant** [comment]
Same test failure at e3bcc92, even though I changed the expected value from `"value"` to `TRUE`.

On my laptop, running the tests using `devtools::test`, 1c4b99f fails this test while e3bcc92 passes all tests.

**Comment by peterhurford** [comment]
@abelcastilloavant It might be a tricky problem related to `R CMD CHECK` (which is different from `devtools::test`). You can run `R CMD CHECK --no-manual .` within the `s3mpi` directory. Then read `..Rcheck/tests/testthat.Rout.fail` to see why the tests failed.

`rr test` from [rrzsh](github.com/peterhurford/rrzsh) also more closely mirrors `R CMD CHECK` (because it is run in the non-interactive environment) so that could help, but it's not as perfect.

**Comment by peterhurford** [comment]
The current master of s3mpi passes on my local but https://github.com/robertzk/s3mpi/commit/e3bcc927f347bf7e548ac26e9cb8546083edf217 results in  

```
~/dev/s3mpi ⌚ 10:26:05 $ rr test
Warning message:
package ‘devtools’ was built under R version 3.2.4
Loading s3mpi
Loading required package: testthat
Testing s3mpi
s3.get:
s3path:
s3read: ....1........
s3store: ..........

Failed -------------------------------------------------------------------------
1. Failure: if cache is TRUE, it does not error --------------------------------
s3read("key", cache = TRUE, serialize = FALSE) not equal to TRUE.
Types not compatible: character vs logical


DONE ===========================================================================
Warning message:
package ‘testthat’ was built under R version 3.2.5
```

...for me.

**Comment by abelcastilloavant** [comment]
Got reproducible behavior. If I run the tests from the s3mpi directory, I get the test failures you get. I've been running them in an R session opened from a different folder before, and I was getting the inverted failure.

**Comment by abelcastilloavant** [comment]
The folder I was working in before created a session that wrote into some `s3mpi` options.

I wrapped tests in `withr::with_options` to make them more state-independent.

File: R/s3read.r [hunk 3]
```diff
@@ -29,18 +29,18 @@
...
-  # to know that, so let's add a slash for them if they forget.
-  last_character <- function(str) {
-    substr(str, nchar(str), nchar(str))
-  }
-  if (last_character(path) != "/") { path <- paste0(path, "/") }
+  path <- add_ending_slash(path)
 
   s3key <- paste(path, name, sep = "")
+  value <- list()
+  dput(value)
```
Comment (peterhurford) [inline R/s3read.r]: I don't think you want these statements anymore.

File: .travis.yml [hunk 4]
```diff
@@ -9,7 +9,7 @@ install:
 - "./travis-tool.sh install_r stringr"
 - "./travis-tool.sh install_r testthat"
 - "./travis-tool.sh install_github jimhester/covr robertzk/testthatsomemore avantcredit/AWS.tools"
-- "./travis-tool.sh install_github kirillseva/cacher"
+- "./travis-tool.sh install_github kirillseva/cacher jimhester/withr"
```
Comment (peterhurford) [inline .travis.yml]: I think this would be covered by `./travis-tool.sh install_deps`. We could also get a more modern .travis.yml.

**Comment by abelcastilloavant** [comment]
Travis failing due to 4 notes:
```
N  checking DESCRIPTION meta-information
   License components which are templates and need '+ file LICENSE':
     MIT
```
```
N  checking top-level files
   File
     LICENSE
   is not mentioned in the DESCRIPTION file.
```
```
N  checking dependencies in R code
   Unexported object imported by a ':::' call: ‘AWS.tools:::check.bucket’
     See the note in ?`:::` about the use of this operator.
```
```
N  checking R code for possible problems
   s3delete: no visible binding for global variable ‘key’
```
```

```

**Comment by peterhurford** [comment]
@abelcastilloavant Nice. Travis doesn't fail because of notes -- it only fails because of warnings or errors.

This failure says the `s3delete` roxygen docs are incomplete:

```
W  checking for code/documentation mismatches
   Codoc mismatches from documentation object 's3delete':
   s3delete
     Code: function(key, path = s3path())
     Docs: function(name, path = s3path())
     Argument names in code not in docs:
       key
     Argument names in docs not in code:
       name
     Mismatches in argument names:
       Position: 1 Code: key Docs: name
```

**Comment by peterhurford** [comment]
There's also a bunch of lintr errors that fail your build. You might want to either remove lintr from the .travis.yml or copy over [the `.lintr` file from batchman](https://github.com/peterhurford/batchman/blob/master/.lintr) for a saner lintr build.

**Comment by abelcastilloavant** [comment]
Good idea, and good idea. 👍

**Comment by abelcastilloavant** [comment]
Removed lintr altogether - no warnings, no errors, but still not passing Travis.

**Comment by peterhurford** [comment]
@abelcastilloavant You also have to remove `checkr::package_exports_checked`... sorry :/

**Comment by abelcastilloavant** [comment]
np @peterhurford - removed it, and Travis now passes.

**Comment by peterhurford** [comment]
LGTM :+1: :shipit:

**Pull request merged**